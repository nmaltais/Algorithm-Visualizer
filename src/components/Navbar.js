/* eslint-disable react/prop-types */
import React, { Component } from 'react';
import {
  Menu,
  Sidebar,
  Button,
  Transition,
  Icon,
} from 'semantic-ui-react';
import { NavLink } from 'react-router-dom';
import './Navbar.scss';

class Navbar extends Component {
  constructor(props) {
    super(props);
    this.state = { showOpenSidebarBtn: false };
  }

  toggleOpenSidebarBtn = () => {
    this.setState((prevState) => ({ showOpenSidebarBtn: !prevState.showOpenSidebarBtn }));
  };

  render() {
    const { showOpenSidebarBtn } = this.state;
    const { toggleSidebar, visible } = this.props;

    const Nav = (props) => (
      <NavLink
        exact
        to={props.to}
        onClick={props.onClick}
        className={props.className}
        activeClassName="active"
      >
        {props.children}
      </NavLink>
    );

    return (
      <>
        <Transition visible={showOpenSidebarBtn} animation="fade right" duration={700}>
          <Button
            icon
            id="openSidebarBtn"
            onClick={() => toggleSidebar('push')}
          >
            <Icon name="bars" />
          </Button>
        </Transition>

        <Sidebar
          as={Menu}
          inverted
          animation="push"
          direction="left"
          vertical
          width="thin"
          visible={visible}
          onHidden={this.toggleOpenSidebarBtn}
          onVisible={this.toggleOpenSidebarBtn}
          style={{ zIndex: 20 }}
        >
          <Menu.Item>
            TOPICS
            <span
              id="closeSidebarBtn"
              role="button"
              tabIndex={0}
              style={{ float: 'right' }}
              onClick={() => toggleSidebar('push')}
              onKeyUp={() => toggleSidebar('push')}
            >
              X
            </span>
          </Menu.Item>
          <Menu.Item>
            <Menu.Header>Sorting</Menu.Header>
            <Menu.Menu>
              <Menu.Item as={Nav} to="/sorting/bubble-sort" name="Bubble Sort" />
              <Menu.Item as={Nav} to="/sorting/selection-sort" name="Selection Sort" />
              <Menu.Item as={Nav} to="/sorting/insertion-sort" name="Insertion Sort" />
              <Menu.Item as={Nav} to="/sorting/quick-sort" name="Quick Sort" />
              <Menu.Item as={Nav} to="/sorting/merge-sort" name="Merge Sort" />
            </Menu.Menu>
          </Menu.Item>

          <Menu.Item>
            <Menu.Header>Pathfinding</Menu.Header>
            <Menu.Menu>
              <Menu.Item as={Nav} to="/pathfinding/BFS" name="BFS" />
              {/* <Menu.Item as={Nav} to="/pathfinding/a_star" name="A Star" /> */}
            </Menu.Menu>
          </Menu.Item>
          {/* <Menu.Item>
            <Menu.Header>Machine Learning</Menu.Header>
            <Menu.Menu>
              <Menu.Item as={Nav} to="/machinelearning/linear-regression" name="Linear Regression" />
            </Menu.Menu>
          </Menu.Item> */}
        </Sidebar>
      </>
    );
  }
}

export default Navbar;
