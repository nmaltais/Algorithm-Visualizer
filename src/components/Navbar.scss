#openSidebarBtn {
  background-color: rgb(27, 28, 29);
  color: white;
  position: fixed; 
  left: 0;
  top: 10px;
  border-radius: 0px 4px 4px 0px;
  z-index: 10,
}