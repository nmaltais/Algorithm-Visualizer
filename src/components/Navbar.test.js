import React from 'react';
import { shallow } from 'enzyme';
import Navbar from './Navbar';

it('renders correctly when open', () => {
  const wrapper = shallow(<Navbar visible toggleSidebar={jest.fn()} />);
  expect(wrapper).toMatchSnapshot();
});

it('renders correctly when closed', () => {
  const wrapper = shallow(<Navbar visible={false} toggleSidebar={jest.fn()} />);
  expect(wrapper).toMatchSnapshot();
});

it('calls toggleOpenSidebarBtn() for #openSidebarBtn onClick', () => {
  // Setup
  const toggleSidebarMock = jest.fn();
  const wrapper = shallow(
    <Navbar visible={false} toggleSidebar={toggleSidebarMock} />,
  );
  wrapper.find('#openSidebarBtn').simulate('click');

  expect(toggleSidebarMock).toHaveBeenCalled();
});

it('calls the toggleSidebar prop for #closeSidebarBtn onClick and onKeyUp', () => {
  // Setup
  const toggleSidebarMock = jest.fn();
  const wrapper = shallow(
    <Navbar visible toggleSidebar={toggleSidebarMock} />,
  );
  wrapper.find('#closeSidebarBtn').simulate('click');
  wrapper.find('#closeSidebarBtn').simulate('keyup');

  expect(toggleSidebarMock).toHaveBeenCalledTimes(2);
});

it('toggleOpenSidebarBtn() changes the showOpenSidebarBtn state value', () => {
  const wrapper = shallow(<Navbar visible={false} toggleSidebar={jest.fn()} />);

  expect(wrapper.state('showOpenSidebarBtn')).toEqual(false);
  wrapper.instance().toggleOpenSidebarBtn();
  expect(wrapper.state('showOpenSidebarBtn')).toEqual(true);
  wrapper.instance().toggleOpenSidebarBtn();
  expect(wrapper.state('showOpenSidebarBtn')).toEqual(false);
});
