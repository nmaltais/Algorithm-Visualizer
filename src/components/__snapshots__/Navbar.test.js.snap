// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when closed 1`] = `
<Fragment>
  <Transition
    animation="fade right"
    duration={700}
    mountOnShow={true}
    transitionOnMount={false}
    unmountOnHide={false}
    visible={false}
  >
    <Button
      as="button"
      icon={true}
      id="openSidebarBtn"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "rgb(27, 28, 29)",
          "borderRadius": "0 4px 4px 0px",
          "color": "white",
          "left": "0",
          "position": "fixed",
          "top": "10px",
          "zIndex": 10,
        }
      }
    >
      <Icon
        as="i"
        name="bars"
      />
    </Button>
  </Transition>
  <Sidebar
    animation="push"
    as={[Function]}
    direction="left"
    inverted={true}
    onHidden={[Function]}
    onVisible={[Function]}
    style={
      Object {
        "zIndex": 20,
      }
    }
    target={
      Object {
        "current": Document {
          "location": Location {
            "assign": [Function],
            "hash": "",
            "host": "localhost",
            "hostname": "localhost",
            "href": "http://localhost/",
            "origin": "http://localhost",
            "pathname": "/",
            "port": "",
            "protocol": "http:",
            "reload": [Function],
            "replace": [Function],
            "search": "",
            "toString": [Function],
          },
        },
      }
    }
    vertical={true}
    visible={false}
    width="thin"
  >
    <MenuItem>
      TOPICS
      <span
        id="closeSidebarBtn"
        onClick={[Function]}
        onKeyUp={[Function]}
        role="button"
        style={
          Object {
            "float": "right",
          }
        }
        tabIndex={0}
      >
        X
      </span>
    </MenuItem>
    <MenuItem>
      <MenuHeader>
        Sorting
      </MenuHeader>
      <MenuMenu>
        <MenuItem
          as={[Function]}
          name="Insert Sort"
          to="/sorting/insert-sort"
        />
        <MenuItem
          as={[Function]}
          name="Bubble Sort"
          to="/sorting/bubble-sort"
        />
        <MenuItem
          as={[Function]}
          name="Selection Sort"
          to="/sorting/selection-sort"
        />
        <MenuItem
          as={[Function]}
          name="Merge Sort"
          to="/sorting/merge-sort"
        />
      </MenuMenu>
    </MenuItem>
    <MenuItem>
      <MenuHeader>
        Pathfinding
      </MenuHeader>
      <MenuMenu>
        <MenuItem
          as={[Function]}
          name="DFS"
          to="/pathfinding/DFS"
        />
        <MenuItem
          as={[Function]}
          name="BFS"
          to="/pathfinding/BFS"
        />
        <MenuItem
          as={[Function]}
          name="A Star"
          to="/pathfinding/a-star"
        />
      </MenuMenu>
    </MenuItem>
  </Sidebar>
</Fragment>
`;

exports[`renders correctly when open 1`] = `
<Fragment>
  <Transition
    animation="fade right"
    duration={700}
    mountOnShow={true}
    transitionOnMount={false}
    unmountOnHide={false}
    visible={false}
  >
    <Button
      as="button"
      icon={true}
      id="openSidebarBtn"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "rgb(27, 28, 29)",
          "borderRadius": "0 4px 4px 0px",
          "color": "white",
          "left": "0",
          "position": "fixed",
          "top": "10px",
          "zIndex": 10,
        }
      }
    >
      <Icon
        as="i"
        name="bars"
      />
    </Button>
  </Transition>
  <Sidebar
    animation="push"
    as={[Function]}
    direction="left"
    inverted={true}
    onHidden={[Function]}
    onVisible={[Function]}
    style={
      Object {
        "zIndex": 20,
      }
    }
    target={
      Object {
        "current": Document {
          "location": Location {
            "assign": [Function],
            "hash": "",
            "host": "localhost",
            "hostname": "localhost",
            "href": "http://localhost/",
            "origin": "http://localhost",
            "pathname": "/",
            "port": "",
            "protocol": "http:",
            "reload": [Function],
            "replace": [Function],
            "search": "",
            "toString": [Function],
          },
        },
      }
    }
    vertical={true}
    visible={true}
    width="thin"
  >
    <MenuItem>
      TOPICS
      <span
        id="closeSidebarBtn"
        onClick={[Function]}
        onKeyUp={[Function]}
        role="button"
        style={
          Object {
            "float": "right",
          }
        }
        tabIndex={0}
      >
        X
      </span>
    </MenuItem>
    <MenuItem>
      <MenuHeader>
        Sorting
      </MenuHeader>
      <MenuMenu>
        <MenuItem
          as={[Function]}
          name="Insert Sort"
          to="/sorting/insert-sort"
        />
        <MenuItem
          as={[Function]}
          name="Bubble Sort"
          to="/sorting/bubble-sort"
        />
        <MenuItem
          as={[Function]}
          name="Selection Sort"
          to="/sorting/selection-sort"
        />
        <MenuItem
          as={[Function]}
          name="Merge Sort"
          to="/sorting/merge-sort"
        />
      </MenuMenu>
    </MenuItem>
    <MenuItem>
      <MenuHeader>
        Pathfinding
      </MenuHeader>
      <MenuMenu>
        <MenuItem
          as={[Function]}
          name="DFS"
          to="/pathfinding/DFS"
        />
        <MenuItem
          as={[Function]}
          name="BFS"
          to="/pathfinding/BFS"
        />
        <MenuItem
          as={[Function]}
          name="A Star"
          to="/pathfinding/a-star"
        />
      </MenuMenu>
    </MenuItem>
  </Sidebar>
</Fragment>
`;
